lib/puppetserver/ca/x509_loader.rb

```diff
@@ -0,0 +1,123 @@
+require 'openssl'
+
+module Puppetserver
+  module Ca
+    # Load, validate, and store x509 objects needed by the Puppet Server CA.
+    class X509Loader
+
+      attr_reader :errors, :certs, :key, :crls
+
+      def initialize(bundle_path, key_path, chain_path = nil)
+        @errors = []
+
+        @certs = load_certs(bundle_path)
+        @key = load_key(key_path)
+        @crls = chain_path ? load_crls(chain_path) : []
+
+        validate(@certs, @key, @crls)
+      end
+
+      # Only do as much validation as is possible, assume whoever tried to
+      # load the objects wrote errors about any invalid ones, but that bundle
+      # and chain may be empty arrays and pkey may be nil.
+      def validate(bundle, pkey, chain)
+        if !chain.empty? && !bundle.empty?
+          validate_crl_and_cert(chain.first, bundle.first)
+        end
+
+        if pkey && !bundle.empty?
+          validate_cert_and_key(pkey, bundle.first)
+        end
+
+        unless bundle.empty?
+          validate_full_chain(bundle, chain)
+        end
+      end
+
+      def load_certs(bundle_path)
+        certs, errs = [], []
+
+        bundle_string = File.read(bundle_path)
+        cert_strings = bundle_string.scan(/-----BEGIN CERTIFICATE-----.*?-----END CERTIFICATE-----/m)
+        cert_strings.each do |cert_string|
+          begin
+            certs << OpenSSL::X509::Certificate.new(cert_string)
+          rescue OpenSSL::X509::CertificateError
+            errs << "Could not parse entry:\n#{cert_string}"
+          end
+        end
+
+        if certs.empty?
+          errs << "Could not detect any certs within #{bundle_path}"
+        end
+
+        unless errs.empty?
+          @errors << "Could not parse #{bundle_path}"
+          @errors += errs
+        end
+
+        return certs
+      end
+
+      def load_key(key_path)
+        begin
+          OpenSSL::PKey.read(File.read(key_path))
+        rescue ArgumentError => e
```

Comment:
If key_path file is empty, will this cause an argumentError?

Reply:
Oh that's weird

What version of Ruby are you using?

```
irb(main):011:0> OpenSSL::PKey.read(File.read('foo-key.pem'))
ArgumentError: Could not parse PKey: no start line
        from (irb):11:in `read'
        from (irb):11
        from bin/console:14:in `<main>'
```

Reply:
^ ruby 2.5 (I know I shouldnt be so high).